src/v/utils/tracking_allocator.h

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright 2020 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/core/sstring.hh>
+
+#include <list>
+#include <memory>
+
+namespace ss = seastar;
+
+namespace util {
+
+/// Tracks allocations and deallocations of memory. A mem_tracker can have
+/// child trackers effectively forming an n-ary tree like structure.
+class mem_tracker {
+public:
+    explicit mem_tracker(ss::sstring label)
+      : _label(std::move(label)) {}
+    mem_tracker(mem_tracker&) = delete;
+    mem_tracker(mem_tracker&&) = default;
+    mem_tracker& operator=(mem_tracker&) = delete;
+    mem_tracker& operator=(mem_tracker&&) = delete;
+    ~mem_tracker() = default;
+
+    void allocate(int64_t bytes) { _consumption += bytes; }
+
+    void deallocate(int64_t bytes) { _consumption -= bytes; }
+
+    ss::lw_shared_ptr<mem_tracker> create_child(ss::sstring);
+
+    ss::sstring label() { return _label; }
+
+    /// Consumption of self + all the children rooted with this tracker..
+    int64_t consumption();
+
+    /// Pretty prints the tracker tree rooted at this instance.
+    /// Example:
+    /// |
+    /// root: 32.000bytes
+    /// | |
+    /// | child1: 24.000bytes
+    /// | | |
+    /// | | child11: 0.000bytes
+    /// | | | |
+    /// | | | child111: 0.000bytes
+    /// | |
+    /// | child2: 0.000bytes
+    ss::sstring debug(int level = 0);
+
+private:
+    ss::sstring _label;
```

Comment:
if it turns out that most labels are string literals (like `mem_tracker("label")`), then you could store a string_view here instead.